File: implementations/fastapi/examples/src/payment/application/query/get_refund_reason_insights_handler.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import date

from ...domain.refund_reason_insights_query import RefundReasonInsightsQuery
from .result import RefundReasonCategoryCountResult, RefundReasonInsightsResult


@dataclass
class GetRefundReasonInsightsQuery:
    from_date: date | None = None
    to_date: date | None = None


class GetRefundReasonInsightsHandler:
    """An ops/analytics read model, not a per-owner one — deliberately not scoped by any
    requester_id, since its whole purpose is to surface refund-reason patterns across every
    refund, not one user's. This repo has no separate admin-authorization boundary, so the
    endpoint is exposed behind the same baseline auth as every other endpoint (see
    payment_router.py); a production system would put this behind a dedicated ops/admin role
    instead.
    """

    def __init__(self, insights_query: RefundReasonInsightsQuery) -> None:
        self._insights_query = insights_query

    async def execute(self, query: GetRefundReasonInsightsQuery) -> RefundReasonInsightsResult:
        counts = await self._insights_query.get_insights(query.from_date, query.to_date)
        return RefundReasonInsightsResult(
            counts=[RefundReasonCategoryCountResult(category=c.category, count=c.count) for c in counts],
            total_classified=sum(c.count for c in counts),
        )

File: implementations/go/examples/internal/application/query/get_refund_reason_insights_handler.go
package query

import (
	"context"
	"fmt"
	"time"

	"github.com/example/account-service/internal/domain/payment"
)

// GetRefundReasonInsightsQuery narrows the aggregation to a created_at date
// range — FromDate/ToDate are the zero time.Time when the caller did not
// supply that bound.
type GetRefundReasonInsightsQuery struct {
	FromDate time.Time
	ToDate   time.Time
}

// GetRefundReasonInsightsHandler — an ops/analytics read model, not a
// per-owner one — deliberately not scoped by RequesterID/OwnerID at all,
// since its whole purpose is to surface refund-reason patterns across every
// refund, not one user's (see payment.RefundReasonInsightsQuery's doc
// comment for why this repo's lack of a separate admin-authorization
// boundary makes this endpoint's simplification acceptable).
type GetRefundReasonInsightsHandler struct {
	insights payment.RefundReasonInsightsQuery
}

func NewGetRefundReasonInsightsHandler(insights payment.RefundReasonInsightsQuery) *GetRefundReasonInsightsHandler {
	return &GetRefundReasonInsightsHandler{insights: insights}
}

func (h *GetRefundReasonInsightsHandler) Handle(ctx context.Context, q GetRefundReasonInsightsQuery) (*RefundReasonInsightsResult, error) {
	counts, err := h.insights.FindReasonInsights(ctx, payment.RefundReasonInsightFilter{
		CreatedFrom: q.FromDate,
		CreatedTo:   q.ToDate,
	})
	if err != nil {
		return nil, fmt.Errorf("get refund reason insights: %w", err)
	}

	results := make([]RefundReasonCategoryCountResult, len(counts))
	total := 0
	for i, c := range counts {
		results[i] = RefundReasonCategoryCountResult{Category: string(c.Category), Count: c.Count}
		total += c.Count
	}
	return &RefundReasonInsightsResult{Counts: results, TotalClassified: total}, nil
}
